controllers/gatekeepersync/gatekeeper_constraint_sync.go

```diff
@@ -0,0 +1,434 @@
+		}
+
+		return true
+	})
+
+	// This is true if a policy used to have Gatekeeper constraints but now does not.
+	if len(constraintsToWatch) == 0 {
+		err = r.ConstraintsWatcher.RemoveWatcher(policyObjID)
+		if err != nil {
+			log.Error(err, "Failed to remove the watches on the constraints")
+
+			return reconcile.Result{}, err
+		}
+	} else {
+		err = r.ConstraintsWatcher.AddOrUpdateWatcher(policyObjID, constraintsToWatch...)
+		if err != nil {
+			log.Error(err, "Failed to update the watches on the constraints")
+
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// sendComplianceEvent wraps SendComplianceEvent and only sends an event if it isn't already set in the policy.
+// Additionally, it adjust the compliance and message if the validating webhook is disabled and the constraint's
+// enforcementAction is set to deny.
+func (r *GatekeeperConstraintReconciler) sendComplianceEvent(
+	ctx context.Context,
+	policy *policyv1.Policy,
+	constraint *unstructured.Unstructured,
+	templateIndex int,
+	msg string,
+	compliance policyv1.ComplianceState,
+) error {
+	enforcementAction, err := getEnforcementAction(constraint)
+	if err != nil {
+		err := r.sendComplianceEvent(ctx, policy, constraint, templateIndex, err.Error(), policyv1.NonCompliant)
+		if err != nil {
+			log.Error(err, "Failed to send the compliance event")
+
+			return err
+		}
+	}
+
+	if enforcementAction == "deny" {
+		webhookEnabled, err := r.webhookEnabled(ctx)
+		if err != nil {
+			log.Error(err, "Failed to determine if the Gatekeeper webhook is enabled")
+
+			return err
+		}
+
+		if !webhookEnabled {
+			compliance = policyv1.NonCompliant
+			msg = fmt.Sprintf(
+				"The Gatekeeper validating webhook is disabled but the constraint's spec.enforcementAction is %s. %s",
+				enforcementAction,
+				msg,
+			)
+		}
+	}
+
+	kn := policyKindName{Policy: policy.Name, Kind: constraint.GetKind(), Name: constraint.GetName()}
+	msgSHA1 := sha1.Sum([]byte(msg))
+
+	if existingMsgSHA1, ok := r.lastSentMessages.Load(kn); ok && existingMsgSHA1.([20]byte) == msgSHA1 {
+		// The message was already sent.
+		return nil
+	}
+
+	refreshedPolicy := &policyv1.Policy{}
+
+	err = r.Get(ctx, types.NamespacedName{Namespace: policy.Namespace, Name: policy.Name}, refreshedPolicy)
+	if err != nil {
+		log.Error(err, "Failed to refresh the cached policy. Will use potentially stale policy for history comparison.")
+
+		refreshedPolicy = policy
+	}
+
+	owner := metav1.OwnerReference{
+		APIVersion: refreshedPolicy.APIVersion,
+		Kind:       refreshedPolicy.Kind,
+		Name:       refreshedPolicy.Name,
+		UID:        refreshedPolicy.UID,
+	}
+
+	if len(refreshedPolicy.Status.Details) < templateIndex+1 ||
+		len(refreshedPolicy.Status.Details[templateIndex].History) == 0 ||
+		refreshedPolicy.Status.Details[templateIndex].History[0].Message != fmt.Sprintf("%s; %s", compliance, msg) {
+		err = utils.SendComplianceEvent(ctx, r, constraint, owner, msg, compliance)
```

Comment:
I'll rework this so that if it attempts to send the same status message and it's already in the Policy status, the entry in the lastSentMessages is deleted.